File: core/src/main/java/org/iumotionlabs/hefesto/feature/portinfo/model/NetworkOverview.java
package org.iumotionlabs.hefesto.feature.portinfo.model;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Record representing a complete network overview with statistics.
 */
public record NetworkOverview(
    List<EnrichedPortBinding> bindings,
    Statistics statistics,
    LocalDateTime generatedAt
) {
    /**
     * Network statistics.
     */
    public record Statistics(
        int totalListening,
        int totalEstablished,
        int tcpCount,
        int udpCount,
        int exposedCount,
        int localOnlyCount
    ) {
        /**
         * Creates statistics from a list of bindings.
         */
        public static Statistics from(List<EnrichedPortBinding> bindings) {
            int listening = 0;
            int established = 0;
            int tcp = 0;
            int udp = 0;
            int exposed = 0;
            int localOnly = 0;

            for (EnrichedPortBinding b : bindings) {
                if ("LISTEN".equals(b.state())) {
                    listening++;
                } else if ("ESTABLISHED".equals(b.state()) || "ESTAB".equals(b.state())) {
                    established++;
                }

                if ("TCP".equals(b.protocol())) {
                    tcp++;
                } else if ("UDP".equals(b.protocol())) {
                    udp++;
                }

                if (b.isExposed()) {
                    exposed++;
                } else if (b.isLocalOnly()) {
                    localOnly++;
                }
            }

            return new Statistics(listening, established, tcp, udp, exposed, localOnly);
        }

        /**
         * Returns a formatted summary line.
         */
        public String toSummaryLine() {
            return "Listening: %d | Established: %d | TCP: %d | UDP: %d | Exposed: %d | Local: %d".formatted(
                totalListening, totalEstablished, tcpCount, udpCount, exposedCount, localOnlyCount
            );
        }
    }

    /**
     * Creates a NetworkOverview from a list of enriched bindings.
     */
    public static NetworkOverview from(List<EnrichedPortBinding> bindings) {
        return new NetworkOverview(bindings, Statistics.from(bindings), LocalDateTime.now());
    }

    /**
     * Returns only listening ports.
     */
    public List<EnrichedPortBinding> listeningPorts() {
        return bindings.stream()
            .filter(b -> "LISTEN".equals(b.state()))
            .toList();
    }

    /**
     * Returns only exposed ports (bound to 0.0.0.0 or ::).
     */
    public List<EnrichedPortBinding> exposedPorts() {
        return bindings.stream()
            .filter(EnrichedPortBinding::isExposed)
            .toList();
    }

    /**
     * Returns only local-only ports (bound to 127.0.0.1 or ::1).
     */
    public List<EnrichedPortBinding> localOnlyPorts() {
        return bindings.stream()
            .filter(EnrichedPortBinding::isLocalOnly)
            .toList();
    }

    /**
     * Returns bindings grouped by process name.
     */
    public Map<String, List<EnrichedPortBinding>> byProcessName() {
        return bindings.stream()
            .collect(Collectors.groupingBy(EnrichedPortBinding::processName));
    }

    /**
     * Returns bindings grouped by PID.
     */
    public Map<Long, List<EnrichedPortBinding>> byPid() {
        return bindings.stream()
            .collect(Collectors.groupingBy(EnrichedPortBinding::pid));
    }

    /**
     * Returns bindings grouped by service category.
     */
    public Map<ServiceInfo.ServiceCategory, List<EnrichedPortBinding>> byServiceCategory() {
        return bindings.stream()
            .filter(b -> b.serviceInfo() != null)
            .collect(Collectors.groupingBy(b -> b.serviceInfo().category()));
    }

    /**
     * Returns unique process names.
     */
    public List<String> uniqueProcessNames() {
        return bindings.stream()
            .map(EnrichedPortBinding::processName)
            .distinct()
            .sorted()
            .toList();
    }

    /**
     * Returns unique PIDs.
     */
    public List<Long> uniquePids() {
        return bindings.stream()
            .map(EnrichedPortBinding::pid)
            .distinct()
            .sorted()
            .toList();
    }

    /**
     * Filters bindings by process name (case-insensitive partial match).
     */
    public List<EnrichedPortBinding> filterByProcessName(String name) {
        String lowerName = name.toLowerCase();
        return bindings.stream()
            .filter(b -> b.processName().toLowerCase().contains(lowerName))
            .toList();
    }

    /**
     * Returns count of Docker containers.
     */
    public long dockerContainerCount() {
        return bindings.stream()
            .filter(EnrichedPortBinding::isDocker)
            .count();
    }
}

File: hefesto-domain/src/portinfo/network_overview.rs
use chrono::{DateTime, Utc};
use serde::Serialize;
use std::collections::HashMap;

use super::enriched_port_binding::EnrichedPortBinding;
use super::port_binding::ConnectionState;
use super::service_info::ServiceCategory;

/// Network statistics.
#[derive(Debug, Clone, Serialize)]
pub struct NetworkStatistics {
    pub total_listening: usize,
    pub total_established: usize,
    pub tcp_count: usize,
    pub udp_count: usize,
    pub exposed_count: usize,
    pub local_only_count: usize,
}

impl NetworkStatistics {
    /// Creates statistics from a list of bindings.
    pub fn from_bindings(bindings: &[EnrichedPortBinding]) -> Self {
        let mut listening = 0;
        let mut established = 0;
        let mut tcp = 0;
        let mut udp = 0;
        let mut exposed = 0;
        let mut local_only = 0;

        for b in bindings {
            match b.state() {
                ConnectionState::Listen => listening += 1,
                ConnectionState::Established => established += 1,
                _ => {}
            }

            match b.protocol() {
                super::port_binding::Protocol::Tcp => tcp += 1,
                super::port_binding::Protocol::Udp => udp += 1,
            }

            if b.is_exposed() {
                exposed += 1;
            } else if b.is_local_only() {
                local_only += 1;
            }
        }

        Self {
            total_listening: listening,
            total_established: established,
            tcp_count: tcp,
            udp_count: udp,
            exposed_count: exposed,
            local_only_count: local_only,
        }
    }

    /// Returns a formatted summary line.
    pub fn to_summary_line(&self) -> String {
        format!(
            "Listening: {} | Established: {} | TCP: {} | UDP: {} | Exposed: {} | Local: {}",
            self.total_listening,
            self.total_established,
            self.tcp_count,
            self.udp_count,
            self.exposed_count,
            self.local_only_count
        )
    }
}

/// A complete network overview with statistics.
#[derive(Debug, Clone, Serialize)]
pub struct NetworkOverview {
    pub bindings: Vec<EnrichedPortBinding>,
    pub statistics: NetworkStatistics,
    pub generated_at: DateTime<Utc>,
}

impl NetworkOverview {
    /// Creates a NetworkOverview from a list of enriched bindings.
    pub fn from_bindings(bindings: Vec<EnrichedPortBinding>) -> Self {
        let statistics = NetworkStatistics::from_bindings(&bindings);
        Self {
            bindings,
            statistics,
            generated_at: Utc::now(),
        }
    }

    /// Returns only listening ports.
    pub fn listening_ports(&self) -> Vec<&EnrichedPortBinding> {
        self.bindings
            .iter()
            .filter(|b| matches!(b.state(), ConnectionState::Listen))
            .collect()
    }

    /// Returns only exposed ports (bound to 0.0.0.0 or ::).
    pub fn exposed_ports(&self) -> Vec<&EnrichedPortBinding> {
        self.bindings.iter().filter(|b| b.is_exposed()).collect()
    }

    /// Returns only local-only ports (bound to 127.0.0.1 or ::1).
    pub fn local_only_ports(&self) -> Vec<&EnrichedPortBinding> {
        self.bindings.iter().filter(|b| b.is_local_only()).collect()
    }

    /// Returns bindings grouped by process name.
    pub fn by_process_name(&self) -> HashMap<String, Vec<&EnrichedPortBinding>> {
        let mut map: HashMap<String, Vec<&EnrichedPortBinding>> = HashMap::new();
        for b in &self.bindings {
            map.entry(b.process_name().to_string()).or_default().push(b);
        }
        map
    }

    /// Returns bindings grouped by PID.
    pub fn by_pid(&self) -> HashMap<u32, Vec<&EnrichedPortBinding>> {
        let mut map: HashMap<u32, Vec<&EnrichedPortBinding>> = HashMap::new();
        for b in &self.bindings {
            map.entry(b.pid()).or_default().push(b);
        }
        map
    }

    /// Returns bindings grouped by service category.
    pub fn by_service_category(&self) -> HashMap<ServiceCategory, Vec<&EnrichedPortBinding>> {
        let mut map: HashMap<ServiceCategory, Vec<&EnrichedPortBinding>> = HashMap::new();
        for b in &self.bindings {
            if let Some(ref si) = b.service_info {
                map.entry(si.category.clone()).or_default().push(b);
            }
        }
        map
    }

    /// Returns unique process names.
    pub fn unique_process_names(&self) -> Vec<String> {
        let mut names: Vec<String> = self
            .bindings
            .iter()
            .map(|b| b.process_name().to_string())
            .collect();
        names.sort();
        names.dedup();
        names
    }

    /// Returns unique PIDs.
    pub fn unique_pids(&self) -> Vec<u32> {
        let mut pids: Vec<u32> = self.bindings.iter().map(|b| b.pid()).collect();
        pids.sort();
        pids.dedup();
        pids
    }

    /// Filters bindings by process name (case-insensitive partial match).
    pub fn filter_by_process_name(&self, name: &str) -> Vec<&EnrichedPortBinding> {
        let lower_name = name.to_lowercase();
        self.bindings
            .iter()
            .filter(|b| b.process_name().to_lowercase().contains(&lower_name))
            .collect()
    }

    /// Returns count of Docker containers.
    pub fn docker_container_count(&self) -> usize {
        self.bindings.iter().filter(|b| b.is_docker()).count()
    }
}
